packages/app/src/cli/utilities/developer-platform-client/app-management-client.ts

```diff
@@ -626,29 +626,39 @@ export class AppManagementClient implements DeveloperPlatformClient {
     bundleUrl,
     skipPublish: noRelease,
   }: AppDeployOptions): Promise<AppDeploySchema> {
-    // `name` is from the package.json package name or the directory name, while
-    // the branding module reflects the current specified name in the TOML.
-    // Since it is technically valid to not have a branding module, we will default
-    // to the `name` if no branding module is present.
-    let updatedName = name
-    const brandingModule = appModules?.find((mod) => mod.specificationIdentifier === BrandingSpecIdentifier)
-    if (brandingModule) {
-      updatedName = JSON.parse(brandingModule.config).name
+    let versionInput
+    if (bundleUrl) {
+      versionInput = {
+        sourceUrl: bundleUrl,
+      }
+    } else {
+      // `name` is from the package.json package name or the directory name, while
+      // the branding module reflects the current specified name in the TOML.
+      // Since it is technically valid to not have a branding module, we will default
+      // to the `name` if no branding module is present.
+      let updatedName = name
+      const brandingModule = appModules?.find((mod) => mod.specificationIdentifier === BrandingSpecIdentifier)
+      if (brandingModule) {
+        updatedName = JSON.parse(brandingModule.config).name
+      }
+      versionInput = {
+        source: {
+          name: updatedName,
+          appModules: (appModules ?? []).map((mod) => {
+            return {
```

Comment:
Should we force to always use a bundleURL and not accept `versionInput` with a an appModule array? This way we can be 100% sure that all versions have a manifest and we don't need to support two formats for the API here.

Reply:
We can do it that way, but to me, it seems a bit wasteful to require a separate bundle upload when there are no assets in the app.

Reply:
It could seem that way, but is a more unified API and the code would be simpler in both CLI and backend, maybe is worth the extra resources for that? 🤔

Reply:
The latest server [change](https://github.com/Shopify/shopify/pull/538723) unifies the API a bit more. Instead of a separate source type, it takes the actual manifest JSON when the app version is directly specified. The code in this PR has been updated accordingly. The back end will still have the responsibility of storing the manifest, but I think the API change removes the biggest concerns and is a reasonable compromise.